slice-dst: Add example for SliceWithHeader

MattesWhite opened the pull request:
This PR is the result of the discussion in #69 .

It adds an example for the `slice-dst` crate and demonstrates how a custom `Vec` can be implemented.

MattesWhite committed:
slice-dst: Add example for SliceWithHeader

CAD97 commented on crates/slice-dst/examples/my_vec.rs:
```diff
@@ -0,0 +1,126 @@
+
+/// Our [`Vec`] implementation.
+///
+/// _Note:_ In contrast to [`std::vec::Vec`] this stores its length on the heap.
```
suggestion: use [`erasable::Thin`](https://cad97.github.io/pointer-utils/erasable/struct.Thin.html) and store `Thin<Box<HeapData<T>>>`, so the pointer is thin (single-`usize`-big) and length/capacity are both exclusively on the heap. `SliceWithHeader` stores its (slice) length inline to support erasing. (The eventual custom derive will omit this.)

MattesWhite replied:
I'm a little bit reluctant to introduce `erasable::Thin` here. It addresses another issue and is from another crate. I would prefer to stick to an honorable mention here and put another example in the `erasable` crate.

CAD97 commented on crates/slice-dst/examples/my_vec.rs:
```diff
@@ -0,0 +1,126 @@
+    }
+    fn iter(&self) -> impl Iterator<Item = &T> {
+        self.0.slice.iter().take(self.len()).map(|t| unsafe {
+            // Safe as only the initialized elements are iterated.
```
issue: unfortunately, this isn't sound as written, because anyone can use `RefCast::ref_cast` to convert between `&MySlice` and `&HeapData`, and change the header, changing how many initialized items are believed to exist. Or... it would be unsound if you added a `DerefMut` impl for `MyVec`, anyway. Having an example that's only sound due to a technicality is probably not the best idea 😄

suggestion: there are a few ways to fix this:

- (lazy) note explicitly that access to `&mut MySlice<_>` would be unsound, due to the above;
- (simple) just wrap `[T]` instead of `HeapData<T>` for `MySlice<T>`, as it doesn't need access to any of the other heap data (and is already a fat pointer, though using `Thin` could avoid this);
- (chore) just inline the `transmute` usage that ref-cast abstracts away; or
- (funky) use the priv-in-pub trick discussed in https://github.com/dtolnay/ref-cast/issues/9 to get an effectively private cast.

MattesWhite replied:
Thank you very much for pointing this out. I chose the _chore_ option with a detailed `// SAFETY:` comment. I hope this is okay for you.

CAD97 commented on crates/slice-dst/examples/my_vec.rs:
```diff
@@ -0,0 +1,126 @@
+        let len = self.len();
+        self.0.slice.iter_mut().take(len).for_each(|t| {
+            unsafe {
+                // Safe as only initialized values iterated.
```
nitpick: I prefer the `// SAFETY: {preconditions satisfied}` format for safety comments like these.

CAD97 reviewed (changes requested):
The example looks good modulo one tricky soundness pitfall (inline) that I want to be sure we aren't accidentally leading users into.

CAD97 commented:
Feel free to ignore the clippy warnings CI reported; I fixed them in #72.

MattesWhite committed:
Remove ref-cast from slice-dts example

The implementation of a 'cast trait' opened the opportunity to violate
`MyVec`'s contract, making the implementation unsound.

MattesWhite committed:
Remove Cargo.lock as it is not required for crates

Keeping Cargo.lock in git often causes unnecessary conflicts when
merging branches.
